src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1078,36 +1077,270 @@ protected void workerDone(Worker worker) {
+    }
+
+    /**
+     * Called to check if either queue has anything to run.  This is just if the queues are empty.  
+     * If there are scheduled tasks queued, but not ready to run, this will still return 
+     * {@code false}.
+     * 
+     * @return {@code true} if there are no tasks in either queue
+     */
+    public boolean isQueueEmpty() {
+      return executeQueue.isEmpty() && scheduleQueue.isEmpty();
+    }
+
+    /**
+     * Call to get the total quantity of tasks within both stored queues.  If you can, 
+     * {@link #isQueueEmpty()} is a more efficient call.  This returns the total amount of items 
+     * in both the execute and scheduled queue.  If there are scheduled tasks which are NOT ready 
+     * to run, they will still be included in this total.
+     * 
+     * @return Total quantity of tasks queued
+     */
+    public int queueSize() {
+      return scheduleQueue.size() + executeQueue.size();
     }
 
     @Override
-    public TaskWrapper getNext() throws InterruptedException {
-      TaskWrapper task;
-      /* must lock as same lock for removal to 
-       * ensure that task can be found for removal
-       */
-      synchronized (queueLock) {
-        task = queue.take();
-        task.executing();  // for recurring tasks this will put them back into the queue
+    protected void startupService() {
+      runningThread = threadFactory.newThread(this);
+      if (runningThread.isAlive()) {
+        throw new IllegalThreadStateException();
+      }
+      runningThread.setDaemon(true);
+      runningThread.setName(threadName);
+      runningThread.start();
+    }
+
+    @Override
+    protected void shutdownService() {
+      Thread runningThread = this.runningThread;
+      this.runningThread = null;
+      runningThread.interrupt();
+    }
+    
+    /**
+     * Called when the queue has been updated and we may need to wake up the consumer thread.
+     */
+    protected void handleQueueUpdate() {
+      if (! startIfNotStarted()) {
+        Thread currRunningThread = runningThread;
+        if (currRunningThread != null) {
+          LockSupport.unpark(currRunningThread);
+        }
+      }
+    }
+
+    @Override
+    public void run() {
+      while (runningThread != null) {
+        try {
+          TaskWrapper nextTask = getNextTask();
+          if (nextTask != null) {
+            switch (priority) {
+              case High:
+                runHighPriorityTask(nextTask);
+                break;
+              case Low:
+                runLowPriorityTask(nextTask);
+                break;
+              default:
+                throw new UnsupportedOperationException();
+            }
+          }
+        } catch (InterruptedException e) {
+          stopIfRunning();
+          Thread.currentThread().interrupt();
+        } catch (Throwable t) {
+          ExceptionUtils.handleException(t);
+        }
+      }
+    }
+    
+    protected TaskWrapper getNextTask() throws InterruptedException {
```

Comment:
I think we could do alot of improvement around here.  I dont think its causing much over head but I bet we could get rid of a few checks reduce complexity and keep pretty good consistency on the queues with a little bit more thought put into it.